File: data structures and algorithms/cpp/08_055.cpp
#include <unordered_map>
#include <vector>
#include <string>
#include <set>
#include <algorithm>

struct TrieNode {
    std::unordered_map<char, TrieNode*> children;
    bool isEndOfWord;
    int lastSearched;  // Timestamp for history
    
    TrieNode() : isEndOfWord(false), lastSearched(-1) {}
    
    ~TrieNode() {
        for (auto& pair : children) {
            delete pair.second;
        }
    }
};

class SearchSuggestionSystem {
private:
    TrieNode* dictionaryTrie;
    TrieNode* historyTrie;
    int searchCount;  // Acts as a timestamp for recency
    
    void collectSuggestions(
            TrieNode* root, 
            const std::string& prefix, 
            std::string currentWord,
            std::vector<std::pair<std::string, int>>& histResults,
            std::vector<std::string>& dictResults, 
            bool isHistory) {
        TrieNode* node = root;
        
        // Navigate to the node corresponding to the prefix
        for (char c : prefix) {
            if (node->children.find(c) == 
                node->children.end()) {
                return;  // Prefix not found
            }
            node = node->children[c];
            currentWord += c;
        }
        
        // DFS to collect all words with this prefix
        dfsCollect(node, currentWord, histResults, 
                  dictResults, isHistory);
    }
    
    void dfsCollect(
            TrieNode* node, 
            std::string currentWord,
            std::vector<std::pair<std::string, int>>& histResults,
            std::vector<std::string>& dictResults, 
            bool isHistory) {
        if (node->isEndOfWord) {
            if (isHistory) {
                histResults.push_back(
                    {currentWord, node->lastSearched});
            } else {
                dictResults.push_back(currentWord);
            }
        }
        
        // Process children in alphabetical order
        std::vector<char> sortedChars;
        for (const auto& pair : node->children) {
            sortedChars.push_back(pair.first);
        }
        std::sort(sortedChars.begin(), sortedChars.end());
        
        for (char c : sortedChars) {
            dfsCollect(node->children[c], currentWord + c, 
                      histResults, dictResults, isHistory);
        }
    }
    
public:
    SearchSuggestionSystem(
            const std::vector<std::string>& dictionary = {}) 
        : dictionaryTrie(new TrieNode()), 
          historyTrie(new TrieNode()), 
          searchCount(0) {
        
        // Initialize with dictionary words if provided
        for (const std::string& word : dictionary) {
            addToDictionary(word);
        }
    }
    
    ~SearchSuggestionSystem() {
        delete dictionaryTrie;
        delete historyTrie;
    }
    
    void addToDictionary(const std::string& word) {
        TrieNode* node = dictionaryTrie;
        for (char c : word) {
            if (node->children.find(c) == 
                node->children.end()) {
                node->children[c] = new TrieNode();
            }
            node = node->children[c];
        }
        node->isEndOfWord = true;
    }
    
    void addToHistory(const std::vector<std::string>& words) {
        for (const std::string& word : words) {
            searchCount++;
            TrieNode* node = historyTrie;
            for (char c : word) {
                if (node->children.find(c) == 
                    node->children.end()) {
                    node->children[c] = new TrieNode();
                }
                node = node->children[c];
            }
            node->isEndOfWord = true;
            node->lastSearched = searchCount;  // Update timestamp
            
            // Also ensure it's in the dictionary
            addToDictionary(word);
        }
    }
    
    std::vector<std::string> suggest(
            const std::string& prefix) {
        std::vector<std::string> results;
        
        // First, get all matching words from history
        std::vector<std::pair<std::string, int>> histMatches;
        std::vector<std::string> dictMatches;
        
        collectSuggestions(historyTrie, prefix, "", 
                          histMatches, dictMatches, true);
        collectSuggestions(dictionaryTrie, prefix, "", 
                          histMatches, dictMatches, false);
        
        // Combine results (history first, then dictionary)
        std::set<std::string> seen;
        
        // Sort history matches by recency (descending)
        std::sort(histMatches.begin(), histMatches.end(),
                 [](const std::pair<std::string, int>& a, 
                    const std::pair<std::string, int>& b) {
                     return a.second > b.second;
                 });
        
        for (const auto& pair : histMatches) {
            const std::string& word = pair.first;
            if (seen.find(word) == seen.end()) {
                results.push_back(word);
                seen.insert(word);
            }
        }
        
        // Sort dictionary matches alphabetically
        std::sort(dictMatches.begin(), dictMatches.end());
        
        for (const std::string& word : dictMatches) {
            if (seen.find(word) == seen.end()) {
                results.push_back(word);
                seen.insert(word);
            }
        }
        
        return results;
    }
};

File: data structures and algorithms/python/08_055.py
class TrieNode:
   def __init__(self):
      self.children = {}
      self.is_end_of_word = False
      self.last_searched = -1  # Timestamp for history

class SearchSuggestionSystem:
   def __init__(self, dictionary=None):
      self.dictionary_trie = TrieNode()
      self.history_trie = TrieNode()
      self.search_count = 0  # Acts as timestamp for recency
      
      # Initialize with dictionary words if provided
      if dictionary:
         for word in dictionary:
            self.add_to_dictionary(word)
   
   def add_to_dictionary(self, word):
      node = self.dictionary_trie
      for char in word:
         if char not in node.children:
            node.children[char] = TrieNode()
         node = node.children[char]
      node.is_end_of_word = True
   
   def add_to_history(self, words):
      for word in words:
         self.search_count += 1
         node = self.history_trie
         for char in word:
            if char not in node.children:
               node.children[char] = TrieNode()
            node = node.children[char]
         node.is_end_of_word = True
         node.last_searched = self.search_count  # Update timestamp
         
         # Also ensure it's in the dictionary
         self.add_to_dictionary(word)
   
   def suggest(self, prefix):
      results = []
      
      # First, get all matching words from history
      history_matches = []
      self._collect_suggestions(
         self.history_trie, prefix, "", history_matches, True
      )
      
      # Then get all matching words from dictionary
      dict_matches = []
      self._collect_suggestions(
         self.dictionary_trie, prefix, "", dict_matches, False
      )
      
      # Combine results (history first, then dictionary,
      # without duplicates)
      seen = set()
      # Sort by recency (descending)
      for word, _ in sorted(
            history_matches, key=lambda x: -x[1]
      ):  
         if word not in seen:
            results.append(word)
            seen.add(word)
      
      for word in sorted(dict_matches):  # Sort alphabetically
         if word not in seen:
            results.append(word)
            seen.add(word)
      
      return results
   
   def _collect_suggestions(
         self, root, prefix, current_word, results, is_history
   ):
      node = root
      
      # Navigate to the node corresponding to the prefix
      for char in prefix:
         if char not in node.children:
            return  # Prefix not found
         node = node.children[char]
         current_word += char
      
      # DFS to collect all words with this prefix
      self._dfs_collect(node, current_word, results, is_history)
   
   def _dfs_collect(
         self, node, current_word, results, is_history
   ):
      if node.is_end_of_word:
         if is_history:
            results.append((current_word, node.last_searched))
         else:
            results.append(current_word)
      
      # Sort alphabetically
      for char, child_node in sorted(node.children.items()):
         self._dfs_collect(
            child_node, current_word + char, results, is_history
         )